webapp/src/ts/modals/training-cards/training-cards.component.ts

```diff
@@ -0,0 +1,217 @@
+import { Store } from '@ngrx/store';
+import { combineLatest, Subscription } from 'rxjs';
+
+import { MmModalAbstract } from '@mm-modals/mm-modal/mm-modal';
+import { XmlFormsService } from '@mm-services/xml-forms.service';
+import { EnketoService } from '@mm-services/enketo.service';
+import { Selectors } from '@mm-selectors/index';
+import { GlobalActions } from '@mm-actions/global';
+import { GeolocationService } from '@mm-services/geolocation.service';
+import { TranslateService } from '@mm-services/translate.service';
+import { TelemetryService } from '@mm-services/telemetry.service';
+
+@Component({
+  selector: 'training-cards-modal',
+  templateUrl: './training-cards.component.html'
+})
+export class TrainingCardsComponent extends MmModalAbstract implements OnInit, OnDestroy {
+
+  constructor(
+    bsModalRef: BsModalRef,
+    private ngZone: NgZone,
+    private store: Store,
+    private xmlFormsService: XmlFormsService,
+    private enketoService: EnketoService,
+    private geolocationService: GeolocationService,
+    private translateService: TranslateService,
+    private telemetryService: TelemetryService,
+  ) {
+    super(bsModalRef);
+    this.globalActions = new GlobalActions(this.store);
+  }
+
+  static id = 'training-cards-modal';
+  private geoHandle:any;
+  private globalActions;
+  form;
+  trainingForm;
+  formWrapperId = 'training-cards-form';
+  modalTitleKey = 'training_cards.modal.title';
+  loadingContent;
+  contentError;
+  hideModalFooter;
+  errorTranslationKey
+  enketoError;
+  enketoStatus;
+  enketoSaving;
+  telemetryData;
+  showConfirmExit;
+  subscription: Subscription = new Subscription();
+
+  ngOnInit() {
+    this.telemetryData = { preRender: Date.now() };
+    this.reset();
+    this.subscribeToStore();
+  }
+
+  ngOnDestroy() {
+    this.subscription.unsubscribe();
+    this.geoHandle && this.geoHandle.cancel();
+    // old code checked whether the component is reused before unloading the form
+    // this is because AngularJS created the new "controller" before destroying the old one
+    // in Angular, unless specific RouteReuseStrategies are employed, components are always
+    // destroyed before new ones are created
+    // see https://github.com/angular/angular/blob/10.2.x/packages/router/src/operators/activate_routes.ts#L37
+    // for Angular behavior
+    // see https://github.com/medic/cht-core/issues/2198#issuecomment-210202785 for AngularJS behavior
+    this.enketoService.unload(this.form);
+    this.globalActions.clearEnketoStatus();
+  }
+
+  private loadForm() {
+    return this.ngZone.runOutsideAngular(() => this._loadForm());
+  }
+
+  private async _loadForm() {
+    try {
+      this.loadingContent = true;
+      this.hideModalFooter = true;
+      this.geoHandle && this.geoHandle.cancel();
+      this.geoHandle = this.geolocationService.init();
+      const form = await this.xmlFormsService.get(this.trainingForm);
+      return this.ngZone.run(() => this.renderForm(form));
```

Comment:
What is this returning? The `renderForm` function doesn't return anything, so presumably this doesn't either? I think you can just drop the `return` keyword.